pkg/daemon/daemon.go

```diff
@@ -423,6 +423,107 @@ func PrepareNamespace(target string) error {
 	return nil
 }
 
+// ReexecuteForTargetRoot detects the OS in the host root filesystem, then
+// copies the appropriate binary into `/run/bin/` there, then does a `chroot`
+// and re-executes the new binary.
+func ReexecuteForTargetRoot(target string) error {
+	// Nothing to do in this case
+	if target == "/" {
+		return nil
+	}
+	// Extra check to avoid recursion
+	reexecEnv := "_MCD_DID_REEXEC"
+	if _, ok := os.LookupEnv(reexecEnv); ok {
+		return nil
+	}
+
+	sourceOsVersion, err := osrelease.GetHostRunningOSFromRoot("/")
+	if err != nil {
+		return fmt.Errorf("failed to get source OS: %w", err)
+	}
+
+	targetOsVersion, err := osrelease.GetHostRunningOSFromRoot(target)
+	if err != nil {
+		return fmt.Errorf("failed to get target OS: %w", err)
+	}
+
+	var sourceBinarySuffix string
+	if sourceOsVersion.IsLikeRHEL() && targetOsVersion.IsLikeRHEL() {
+		sourceMajor := sourceOsVersion.BaseVersionMajor()
+		targetMajor := targetOsVersion.BaseVersionMajor()
+		if sourceMajor == "8" && targetMajor == "9" {
```

Comment:
might be contrived but this just popped into my head
```
if sourceMajor != "" {
  sourceBinarySuffix = ".rhel"+sourceMajor
  klog.Info(fmt.Sprintf("targeting rhel%s which is a different version than the container rhel%s",  targetMajor, sourceMajor))
} else {
  klog.Info("not chrooting for source=rhel-%s target=rhel-%s", sourceMajor, targetMajor)
}
```

Reply:
baseversion comes from /etc/os-release which is part of standard package. We can rely that it won't break otherwise we would be breaking multiple consumers of os-release file.